### PR TITLE
Feat/52/Crud for division

### PR DESCRIPTION
**Pull request by NataliaHerus**

### CODE REVIEW

File: MilitaryManager.Core/Services/DivisionService.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+using AutoMapper;
+using MilitaryManager.Core.DTO.Divisions;
+using MilitaryManager.Core.Entities.DivisionEntity;
+using MilitaryManager.Core.Entities.DivivsionEntity;
+using MilitaryManager.Core.Interfaces.Repositories;
+using MilitaryManager.Core.Interfaces.Services;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+
```
Comment (vladtymo) [inline MilitaryManager.Core/Services/DivisionService.cs]: Empty line.

File: MilitaryManager.Core/Services/DivisionService.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        public async Task<IEnumerable<DivisionDTO>> GetAllDivisionsAsync()
+        {
+            Specification<Division> specification = new Divisions.RootDivisionList();
+            var divisions = await _divisionRepository.GetListBySpecAsync(specification);
+
+            return _mapper.Map<IEnumerable<DivisionDTO>>(divisions);
+        }
+
+        public async Task<IEnumerable<DivisionDTO>> GetDivisionByKeyAsync(int id)
```
Comment (vladtymo) [inline MilitaryManager.Core/Services/DivisionService.cs]: Why do not return a single object?

File: MilitaryManager.Core/Services/DivisionService.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            Specification<Division> specification = new Divisions.DivisionsListById(id);
+            var getDivision = await _divisionRepository.GetListBySpecAsync(specification);
+
+            return _mapper.Map<IEnumerable<DivisionDTO>>(getDivision);
+        }
+
+        public async Task<DivisionDTO> CreateDivisionAsync(DivisionDTO query)
```
Comment (vladtymo) [inline MilitaryManager.Core/Services/DivisionService.cs]: Change the parameter name 'query' to more appropriate (dto, data, divisionDto)

File: MilitaryManager.Core/Services/DivisionService.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        public DivisionService(IRepository<Division, int> divisionRepository, IMapper mapper)
+        {
+            _divisionRepository = divisionRepository;
+            _mapper = mapper;
+        }
+        public async Task<IEnumerable<DivisionDTO>> GetAllDivisionsAsync()
+        {
+            Specification<Division> specification = new Divisions.DivisionsList();
+            var divisions = await _divisionRepository.GetListBySpecAsync(specification);
+            var getDivisions = divisions.Where(x => x.ParentId == null);
```
Comment (vladtymo) [inline MilitaryManager.Core/Services/DivisionService.cs:28]: Why did you move it here from the Specification class?

File: MilitaryManager.Core/Services/DivisionService.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            var division = _mapper.Map<Division>(dto);
+            var updateDivision = await _divisionRepository.UpdateAsync(division);
+            await _divisionRepository.SaveChangesAcync();
+
+            return _mapper.Map<DivisionDTO>(updateDivision);
+        }
+
+        public async Task<DivisionDTO> DeleteDivisionAsync(int id)
+        {
+            var division = new Division() { Id = id };
```
Comment (vladtymo) [inline MilitaryManager.Core/Services/DivisionService.cs]: Check if division exists before the delete.

**Review by vladtymo — CHANGES_REQUESTED** [review]
It would be good to add validations for the DTO classes.

**Review by vladtymo — APPROVED** [review]

File: MilitaryManager.Units.API/Controllers/DivisionController.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    public class DivisionController : ControllerBase
+    {
+        protected readonly IDivisionService _divisionService;
+
+        public DivisionController(IDivisionService divisionService)
+        {
+            _divisionService = divisionService;
+        }
+
+        [HttpGet]
```
Comment (Alex-knu) [inline MilitaryManager.Units.API/Controllers/DivisionController.cs:21]: add [Route("collection")]

**Review by Alex-knu — CHANGES_REQUESTED** [review]

File: MilitaryManager.Core/Services/DivisionService.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            var divisions = await _divisionRepository.GetListBySpecAsync(specification);
+            var getDivisions = divisions.Where(x => x.ParentId == null);
+
+            return _mapper.Map<IEnumerable<DivisionDTO>>(getDivisions);
+        }
+
+        public async Task<DivisionDTO> GetDivisionByKeyAsync(int id)
+        {
+            Specification<Division> specification = new Divisions.DivisionsList();
+            var division = await _divisionRepository.GetListBySpecAsync(specification);
```
Comment (Alex-knu) [inline MilitaryManager.Core/Services/DivisionService.cs:36]: Use GetFirstBySpecAsync(specification)

**Review by Alex-knu — CHANGES_REQUESTED** [review]

File: MilitaryManager.Core/Services/DivisionService.cs (hunk 7 above)
Comment (NataliaHerus) [reply MilitaryManager.Core/Services/DivisionService.cs:36]: If I do it in such a way, I won't be able to see the subdivisions of subdivision

**Review by Alex-knu — APPROVED** [review]

**Review by vladtymo — APPROVED** [review]
